zkevm-circuits/src/evm_circuit/util/memory_gadget.rs

```diff
@@ -133,25 +185,181 @@ impl<F: Field> MemoryAddressGadget<F> {
+/// . [memoryGasCost](https://github.com/ethereum/go-ethereum/blob/db18293c32f6dc5d6886e5e68ab8bfd12e33cad6/core/vm/gas_table.go#L38)
+/// . [toWordSize](https://github.com/ethereum/go-ethereum/blob/db18293c32f6dc5d6886e5e68ab8bfd12e33cad6/core/vm/common.go#L67)
+#[derive(Clone, Debug)]
+pub(crate) struct MemoryExpandedAddressGadget<F> {
+    length_is_zero: IsZeroGadget<F>,
+    offset_length_sum: AddWordsGadget<F, 2, false>,
+    sum_lt_cap: LtGadget<F, N_BYTES_U64>,
+    sum_within_u64: IsZeroGadget<F>,
+}
+
+impl<F: Field> CommonMemoryAddressGadget<F> for MemoryExpandedAddressGadget<F> {
+    fn construct_self(cb: &mut EVMConstraintBuilder<F>) -> Self {
+        let offset = cb.query_word32();
+        let length = cb.query_word32();
+        let sum = cb.query_word32();
+
+        let sum_lt_cap = LtGadget::construct(
+            cb,
+            from_bytes::expr(&sum.limbs[..N_BYTES_U64]),
+            (MAX_EXPANDED_MEMORY_ADDRESS + 1).expr(),
+        );
+
+        let sum_overflow_hi = sum::expr(&sum.limbs[N_BYTES_U64..]);
+        let sum_within_u64 = IsZeroGadget::construct(cb, sum_overflow_hi);
+
+        let length_is_zero = IsZeroGadget::construct(cb, sum::expr(&length.limbs));
+        let offset_length_sum = AddWordsGadget::construct(cb, [offset, length], sum);
+
+        Self {
+            length_is_zero,
+            offset_length_sum,
+            sum_lt_cap,
+            sum_within_u64,
+        }
+    }
+
+    fn assign(
+        &self,
+        region: &mut CachedRegion<'_, '_, F>,
+        offset: usize,
+        memory_offset: U256,
+        memory_length: U256,
+    ) -> Result<u64, Error> {
+        let length_bytes = memory_length
+            .to_le_bytes()
+            .iter()
+            .fold(0, |acc, val| acc + u64::from(*val));
+        self.length_is_zero
+            .assign(region, offset, F::from(length_bytes))?;
+
+        let (sum, sum_word_overflow) = memory_offset.overflowing_add(memory_length);
+        self.offset_length_sum
+            .assign(region, offset, [memory_offset, memory_length], sum)?;
+
+        self.sum_lt_cap.assign(
+            region,
+            offset,
+            F::from(sum.low_u64()),
+            F::from(MAX_EXPANDED_MEMORY_ADDRESS + 1),
+        )?;
+
+        let sum_overflow_hi_bytes = sum.to_le_bytes()[N_BYTES_U64..]
+            .iter()
+            .fold(0, |acc, val| acc + u64::from(*val));
+        self.sum_within_u64
+            .assign(region, offset, F::from(sum_overflow_hi_bytes))?;
+
+        let address = if length_bytes == 0
+            || sum_overflow_hi_bytes != 0
+            || sum_word_overflow
+            || sum.low_u64() > MAX_EXPANDED_MEMORY_ADDRESS
+        {
+            0
+        } else {
+            sum.low_u64()
+        };
+
+        Ok(address)
+    }
+
+    fn offset_word(&self) -> Word<Expression<F>> {
+        let addends = self.offset_length_sum.addends();
+        // addends[0].expr()
```

Comment:
a leftover

Reply:
removed in https://github.com/privacy-scaling-explorations/zkevm-circuits/pull/1558/commits/0092e00ed175e52fae7541e16576daf26057a769 thanks!